src/picongpu/include/algorithms/KinEnergy.hpp

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright 2017 Heiko Burau
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "simulation_defines.hpp"
+#include "algorithms/Gamma.hpp"
+
+namespace picongpu
+{
+
+using namespace PMacc;
+
+template<typename precisionType = float_X>
+struct KinEnergy
+{
+    typedef precisionType valueType;
+
+    template<typename MomType, typename MassType >
+    HDINLINE valueType operator()(const MomType mom, const MassType mass)
+    {
+        if(mass == MassType(0.0))
+        {
+            return SPEED_OF_LIGHT * math::abs(precisionCast<valueType>(mom));
```

Comment:
sure, but is your momentum already scaled correctly to Planck constant / lambda? :)